### ISSUE TITLE
CUDA Installation difficulty 

### ISSUE THREAD

**Issue by madisonpb1** [lmb-freiburg/Unet-Segmentation#74]
Disclaimer: I am a biophysicist/beginner coder and this has surpassed my troubleshooting abilities. I was hired (after someone had built the gaming computer) to install this UNet software and mainly work on the frontend (cell counting). However, establishing the backend has been a sluggish struggle and I would greatly appreciate any assistance. (I've referenced several previously posted issues with no luck at resolution). 

I'm working on with an Aorus GPU partitioned with windows 10 and ubuntu (20.04). I have no experience with partitions so I'm learning as I go. I assumed I could use the pre-built binaries to set up the server. 

I've managed to use the tests "caffe" and "ssh [hostname] caffe" and receive usage messages in response. I believe I've successfully installed the U-Net package. I've also located the bash file (with the command "xdc-open ~/.bashrc"), and added the appropriate library paths at the top as indicated in the instructions.  

However, I keep receiving error messages when attempting to install the CUDA library. 
Should I opt for a later version than 8.0 since I'm running Ubuntu 20.04? 
Would I need to build the CUDA myself and if so, how? 

The two error messages I received when working on this step (https://lmb.informatik.uni-freiburg.de/resources/opensource/unet/#installation-backend-onsite) are as follows. 
The line "sudo apt-get update" leads to this error message: 
"The following signatures were invalid [First & Last Name] (email address)."
The line "sudo apt-get install -y cuda-8.0 unzip" leads to this error message: 
"Unable to correct problems, you have held broken packages"

Please see the attached pdf screenshots of my terminal for more info: 
[CUDA Difficulty 2.pdf](https://github.com/lmb-freiburg/Unet-Segmentation/files/5466943/CUDA.Difficulty.2.pdf)
[CUDA Difficulty 1.pdf](https://github.com/lmb-freiburg/Unet-Segmentation/files/5466944/CUDA.Difficulty.1.pdf)

**Comment by ThorstenFalk** [comment]
Hmz, you want to install cuda from an ubuntu 16.04 repository. That's probably not the best idea when running Ubuntu 20.04. There is no pre-built package for Ubuntu 20.04, so I would opt for installing from source using the latest nVidia driver and a more recent cuda version.

The error message "The following signatures were invalid [First & Last Name] (email address)." is a problem with your apt cache and your apt sources. This is not related to caffe_unet and probably has to be fixed before you can install anything on the computer. After resolving the general apt-problems on your machine you can continue using the standard Ubuntu packages:

Ubuntu 20.04 has a cuda package in its standard repositories:
```
$ sudo apt update
$ sudo apt install nvidia-cuda-toolkit
```
You probably must remove the incomplete cuda 8 installation beforehand.

Then follow "Setup on own server (from source)". Maybe package names changed in Ubuntu 20.04, so if there is no installation candidate for some of the packages try to find the current name using "apt-cache search". For your GPU you will probably also need to extend the cuda compute capabilities to 7.5 for your Turing GPU. This would change the cmake configuration to
```
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/home/unetuser/u-net -DUSE_OPENCV=OFF -DUSE_LEVELDB=OFF -DUSE_LMDB=OFF -DBUILD_python=OFF -DBUILD_python_layer=OFF -DCUDA_ARCH_NAME=Manual -DCUDA_ARCH_BIN="50 60 61 70 72 75" -DCUDA_ARCH_PTX="35" ..
```

However I never tried the backend on a more recent CC version than 61, so caffe may not support it at all. I anyways wish you good luck!

**Comment by ThorstenFalk** [comment]
Ah, and in any case post your experience, that you and others can look it up!

**Comment by madisonpb1** [comment]
Since this posting, I resolved the general apt-problems with key updates and necessary installations on the machine. I also verified the CUDA toolkit is installed (v 10.1). Thank you, Thorsten for the suggestions. 

My biggest challenge remains the cmake command. I am having immense difficulty addressing the recurring errors. 
![CMake Command Output 1- Argument error](https://user-images.githubusercontent.com/73542919/98687061-ac909380-233f-11eb-9899-32e4d41483e9.png)
![CMake Command Output 2- cuDNN Not Found](https://user-images.githubusercontent.com/73542919/98687063-ad292a00-233f-11eb-9dba-83aa7472bbbd.png)

At first, the cmake command returned a "Could NOT find Doxygen (missing: DOXYGEN_EECUTABLE)" 
I was able to address this with the apt install of doxygen. 

An error I have been unable to address is "CMake Error at src/caffe/CMakeLists.txt:30 (target_compile_definitions):
target_compile_definitions called with incorrect number of arguments"

[CMakeLists.txt](https://github.com/lmb-freiburg/Unet-Segmentation/files/5517928/CMakeLists.txt)

Another perplexing issue is under the NVIDIA CUDA header on the output of CMake, the cuDNN reads "Not Found". Regardless, the CMake command consistently fails when I try to run it. 

![CMake Command fails](https://user-images.githubusercontent.com/73542919/98687059-abf7fd00-233f-11eb-9786-3bd5864c8326.png)

I suspect my bash folder is failing to accurately establish pathways of communication but I am uncertain how to proceed. 
These lines are contained in the .bashrc file:
```
#setting the envrionment for UNET cell count ImageJ plugin
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/minions/local/cuda-8.0/lib64/
#setting the environment for the caffe UNET software
export PATH=$PATH:/home/unetuser/unet/bin
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/unetuser/unet/lib:home/unetuser/unet/extlib
```
and 

```
export LD_LIBRARY_PATH=/usr/lib/cuda/lib64:$LD_LIBRARY_PATH
export LD_LIBRARY_PATH=/usr/lib/cuda/include:$LD_LIBRARY_PATH
export LD_LIBRARY_PATH=/usr/lib/cuda/lib64:$LD_LIBRARY_PATH
export LD_LIBRARY_PATH=/usr/lib/cuda/include:$LD_LIBRARY_PATH

```

**Comment by ThorstenFalk** [comment]
Good news that you resoved the apt problems. The remaining problem is not 100% clear, it's definitely not a problem with your environment variables.

1. doxygen does not matter, it's only for generating documentation you won't need, so you could simply disable build docs to get rid of this problem, but installing doxygen doesn't harm
2. You only installed the cuda toolkit. cuDNN is on top and must be explicitly downloaded from the nVidia website (which I highly recommend). If it is still as it was a year ago you will have to register as nVidia developer to do so, but it's for free, no spam, or whatsoever.
3. The actual error is that the cmake variable Caffe_Definitions is not set when used in src/caffe/CMakelists.txt. However the reason is unclear to me, you can try to simply comment out line 30, but the fact the the variable is not set when it should be is no good sign.

**Comment by madisonpb1** [comment]
Yes, I had downloaded the cuDNN from the nVidia developer website prior to posting my updated issue. I extracted the files into the caffe folder but the errors persist. 
Could you explain where the cuDNN files should be extracted so the CUDA Toolkit is effectively communicating? Do I need to establish pathways for the cuDNN package? 

As for line 30 in the CMakeLists.txt, I changed it as the comments suggested from:
`caffe_option(USE_CUDA  "Build Caffe without CUDA support" OFF) 
# TODO: renamed from CPU_ONLY to USE_CUDA on Nov 12, 2020`

**Comment by ThorstenFalk** [comment]
cuDNN should be ideally placed in the cuda/lib64 folder next to libcudart.so then it should be automatically found, but you can place it wherever you want if specifying the location during cmake configuration. At best you use ccmake (apt install cmake-curses-gui) for configuration, then you can interactively configure all aspects of caffe_unet.

You commented out line 30 of caffe_unet/CMakeLists.txt, but the error refers to line 30 of caffe_unet/src/caffe/CMakeLists.txt. Revert the root CMakeLists.txt and edit src/caffe/CMakeLists.txt instead, the line should currently be:
```
target_compile_definitions(caffe ${Caffe_DEFINITIONS})
```

**Comment by madisonpb1** [comment]
After troubleshooting for some time, I felt compelled to downgrade the Ubuntu version from 20.04 to 16.04. I'm using the pre-built binaries to finish the backend server setup.  I had hoped this would make the U-Net installation a lot more approachable. 

However, I am encountering a new-to-me error with caffe (similar to Issue #54 and #68) but I'm struggling to find a solution. 
Given the command `caffe`
the output is: `caffe: error while loading shared libraries: libglog.so.0: cannot open shared object file: No such file or directory`

Now that I'm working in the Windows10/Ubuntu16.04 dual environment, I'm having trouble identifying my library deficiencies. How might I fix my shared libraries error? 

More info: 
`Which caffe` 
`/home/minions/unetuser/u-net/bin/caffe`

`echo $LD_LIBRARY_PATH`
`:usr/local/cuda-8.0/lib64/:/home/minions/unetuser/u-net/lib:/home/unetuser/u-net/extlib`

**Comment by madisonpb1** [comment]
I've resolved the library issues by running `export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/unetuser/u-net/lib:/home/unetuser/u-net/extlib` in the terminal rather than the bash file per the recommendation of another. 
Now Caffe works! 

However, I cannot get ssh to work. 
I downloaded PuTTY on the Windows frontend thinking that would help establish a server connection. It didn't. 
I thought perhaps I have too many IP addresses so I tried to locate which one is correct. This computer is on a university network. 

When I input: `ssh [hostname]@[IPv4Address] caffe` into the Ubuntu 16.04 terminal, the output is: `Unable to negotiate with [IPv4Address] port 22: no matching host key type found. Their offer: ssh-dss`

Alternatively, when I input: `ssh [hostname]@[DefaultGateway] caffe` into the Ubuntu 16.04 terminal, the output is: `ssh: connect to host [DefaultGateway] port 22: No route to host`

Is there any explanation for the last step? I have scoured the repositories and forums for similar issues with solutions and no luck. Please, please help me.

**Comment by ThorstenFalk** [comment]
When putting the LD_LIBRARY_PATH statements to .bashrc you must make sure that .bashrc is actually executed, which usually requires to additionally put a line
```
. ~/.bashrc
```
into .profile, then the variable should be set. You can verify using
```
echo $LD_LIBRARY_PATH
```
in a terminal. Explicitly setting the variables before starting caffe_unet works, that's good, but the plugin spawns a new shell, thus you cannot do this when calling caffe_unet via the Fiji plugin. There is a more or less undocumented extra-line in the plugin configuration panel that allows to run user-defined server commands before the actual caffe_unet call, so you could use that if everything else fails.

Is the first try an ssh to localhost (via its global IP)? If so, I don't really get it. I already had issues with an old version of jsch and a very new ssh server which used different key types, but ssh server and client on the Ubuntu 16.04 host should definitely be able to talk to each other. The second error simply indicates that there is no such computer in your network with enabled ssh server. The placeholders are quite puzzling anyways, why do you connect to different machines, and why would you try to connect to the DefaultGateway via ssh? The gateway should never appear in an ssh command, only the end point.

1. Try in a terminal directly on your Ubuntu 16.04 unet server:
```
ssh unetuser@localhost caffe_unet
```
If this works, your server is in principle setup correctly including LD_LIBRARY_PATH.

2. Try to connect using putty from your Windows machine (check the IP using ifconfig in a terminal on the Ubuntu 16.04 machine just to be sure) and try to run caffe_unet in the terminal

3. Try with the plugin. I see no reason why it should not work if 1 and 2 do.

**Comment by madisonpb1** [comment]
The IP address confusion was my fault (I've since identified the correct hostname and respective IP address). 
And the bash code lines are in the bash file as instructed. 

Step 1 does not work in my Ubuntu terminal. I'm guessing my server is not set up correctly but I'm at a total loss on what could be wrong. 
I've troubleshooted this SSH connection as much as I can. I've checked firewalls, OpenSSH is installed on both the Windows 10 and Ubuntu 16 side, the port 22 is active, the SSH is running. I used powershell on windows as admin to make the connection and failed. I also tried using PuTTY on Windows to make SSH connection between the two, no success. My error is consistently a network time out (https://www.ssh.com/ssh/putty/putty-manuals/0.68/Chapter10.html#errors-conntimedout).  
I even tried to change the time out time on PuTTY so I could establish the SSH connection (https://askubuntu.com/questions/254750/how-to-make-putty-ssh-connection-never-to-timeout-when-user-is-idle). 

I cannot figure out why I'm unable to establish the connection. 
And then I ran into this: https://askubuntu.com/questions/1115996/access-ubuntu-terminal-in-window-10-in-dual-booting-system
I'm afraid I'll have to close this issue and never resolve it. All indications on the internet forums I've read say such a connection is not possible on the same machine.
